src/Provider/Apple.php

```diff
@@ -61,7 +61,29 @@ protected function initialize()
         $this->AuthorizeUrlParameters['response_mode'] = 'form_post';
     }
 
+    /**
+     * {@inheritdoc}
+     */
+    protected function getAuthorizeUrl($parameters = [])
+    {
+        $this->AuthorizeUrlParameters = !empty($parameters)
+            ? $parameters
+            : array_replace(
+                (array) $this->AuthorizeUrlParameters,
+                (array) $this->config->get('authorize_url_parameters')
+            );
+
+        if ($this->supportRequestState) {
+            if (!isset($this->AuthorizeUrlParameters['state'])) {
+                $this->AuthorizeUrlParameters['state'] = 'HA-' . str_shuffle('ABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890');
+            }
+
+            $this->storeData('authorization_state', $this->AuthorizeUrlParameters['state']);
+        }
 
+        return $this->authorizeUrl . '?' . http_build_query($this->AuthorizeUrlParameters, '', '&', PHP_QUERY_RFC3986);
+    }
```

Comment:
Why not simply extend OAUTH2.php instead? https://github.com/sputnik73/hybridauth/commit/9909c40a8a5d2b4b7c7ec42e8b8c8baf62233e2d

Reply:
@sputnik73
you changed main class? i just override the method to avoid change the main oauth2 class.

Reply:
I understand we should avoid changing main classes. However it would prevent 20 lines of copy & paste and maybe justify an additional generic setting - I am not sure what's best practice?